File: 2015/python/08.py
# Advent of Code 2015, day 8
# (c) aichingert

def part1(inp):
    c = 0
    ch = 0
    for l in inp:
        c += len(l)
        i = 0
        while i < len(l)-1:
            if l[i] == "\\":
                if l[i+1] == "\\" or l[i+1] == "\"":
                    i += 1
                    ch += 1
                elif l[i+1] == "x":
                    i += 3
                    ch += 1
            elif l[i] != "\"":
                ch += 1
            i+=1
    return c - ch

def part2(inp):
    c = 0
    n = 0
    for l in inp:
        c += len(l)
        for i in range(len(l)):
            if l[i] == "\"" or l[i] == "\\":
                n += 2
            else:
                n += 1
        n += 2
    return n - c

inp = open('../input/08').read().split('\n')
print("Part 1:",part1(inp))
print("Part 2:",part2(inp))

File: 2015/rust/08.rs
// Advent of Code 2015, day 8
// (c) aichingert

fn part1(inp: &str) -> u32 {
    inp.lines()
        .map(|l| {
            let ch = l.chars().collect::<Vec<char>>();
            let mut i: usize = 0;
            let mut c: u32 = 0;

            while i < ch.len()-1 {
                match ch[i] {
                    '\\' => match ch[i+1] {
                        '\\' | '"' => { i+=1; c+=1; },
                        'x' => { i+=3; c+=1 },
                        _ => panic!("invalid input"),
                    },
                    '"' => (),
                    _ => c+=1,
                };
                i += 1;
            }

            ch.len() as u32 - c
        }).sum::<u32>() 
}

fn part2(inp: &str) -> u32 {
    inp.lines()
        .map(|l| {
            let ch = l.chars().collect::<Vec<char>>();
            ch.iter()
                .map(|e| match e {
                    '"' | '\\' => 2,
                    _ => 1,
                }).sum::<u32>() + 2 - ch.len() as u32
        }).sum::<u32>()
}

fn main() {
    let inp = std::fs::read_to_string("../input/08").unwrap();
    println!("Part 1: {:?}", part1(&inp));
    println!("Part 2: {:?}", part2(&inp));
}
